pkg/controller/beat/common/config.go

```diff
@@ -13,66 +13,92 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/settings"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
 
-// setOutput will set the output section in Beat config according to the association configuration.
-func setOutput(cfg *settings.CanonicalConfig, client k8s.Client, associated commonv1.Association) error {
+type DefaultConfig struct {
+	// Managed config is handled by ECK and will not be nullified by user provided config.
+	Managed *settings.CanonicalConfig
+	// Unmanaged config is default config that will be overridden by any user provided config.
+	Unmanaged *settings.CanonicalConfig
+}
+
+// buildOutputConfig will create the output section in Beat config according to the association configuration.
+func buildOutputConfig(client k8s.Client, associated beatv1beta1.BeatESAssociation) (*settings.CanonicalConfig, error) {
 	if !associated.AssociationConf().IsConfigured() {
-		return nil
+		return settings.NewCanonicalConfig(), nil
 	}
 
-	username, password, err := association.ElasticsearchAuthSettings(client, associated)
+	username, password, err := association.ElasticsearchAuthSettings(client, &associated)
 	if err != nil {
-		return err
+		return settings.NewCanonicalConfig(), err
 	}
 
-	esOutput := settings.MustCanonicalConfig(
-		map[string]interface{}{
-			"output.elasticsearch": map[string]interface{}{
-				"hosts":    []string{associated.AssociationConf().GetURL()},
-				"username": username,
-				"password": password,
-			},
-		})
-
-	if err := cfg.MergeWith(esOutput); err != nil {
-		return err
+	esOutput := map[string]interface{}{
+		"output.elasticsearch": map[string]interface{}{
+			"hosts":    []string{associated.AssociationConf().GetURL()},
+			"username": username,
+			"password": password,
+		},
 	}
 
 	if associated.AssociationConf().GetCACertProvided() {
-		if err := cfg.MergeWith(settings.MustCanonicalConfig(
-			map[string]interface{}{
-				"output.elasticsearch.ssl.certificate_authorities": path.Join(CAMountPath, CAFileName),
-			})); err != nil {
-			return err
-		}
+		esOutput["output.elasticsearch.ssl.certificate_authorities"] = []string{path.Join(certificatesDir(&associated), CAFileName)}
 	}
 
-	return nil
+	return settings.NewCanonicalConfigFrom(esOutput)
+}
+
+// BuildKibanaConfig builds on optional Kibana configuration for dashboard setup and visualizations.
+func BuildKibanaConfig(client k8s.Client, associated beatv1beta1.BeatKibanaAssociation) (*settings.CanonicalConfig, error) {
+	if !associated.AssociationConf().IsConfigured() {
+		return settings.NewCanonicalConfig(), nil
+	}
+
+	username, password, err := association.ElasticsearchAuthSettings(client, &associated)
+	if err != nil {
+		return settings.NewCanonicalConfig(), err
+	}
+
+	kibanaCfg := map[string]interface{}{
+		"setup.dashboards.enabled": true,
+		"setup.kibana": map[string]interface{}{
+			"host":     associated.AssociationConf().GetURL(),
+			"username": username,
+			"password": password,
+		},
+	}
```

Comment:
Should we consider adding`setup.dashboards.retry.enabled` ([docs](https://www.elastic.co/guide/en/beats/filebeat/7.0/configuration-dashboards.html#_setup_dashboards_retry_enabled)) to avoid crashing?

Reply:
Not sure: it would hide this unpleasant startup behaviour but it could hide an actual misconfiguration by never failing. Having said that given the low priority of the auto-install dashboards feature I am happy to include it.

Reply:
That's true, but with this change it will be a common behavior between ES and Kibana - if someone configures output incorrectly Beat won't fail either. We can try to come up with a common solution as a part of https://github.com/elastic/cloud-on-k8s/issues/3197.

Reply:
Turns out there is a bug in libbeat that prevents the retry from actually working https://github.com/elastic/beats/issues/18217 because a Kibana Client is created before the retry logic starts and does a first request that fails.

https://github.com/elastic/beats/blob/0c15394a85134e1797d30592cd08789c4c1a8e1d/libbeat/cmd/instance/beat.go#L782